zkvm/src/ops.rs

```diff
@@ -277,3 +279,213 @@ impl Instruction {
         }
     }
 }
+
+impl Program {
+    pub fn new() -> Self {
+        Program(vec![])
+    }
+
+    pub fn to_vec(&self) -> Vec<Instruction> {
```

Comment:
let's consume `self` here and avoid cloning a vector. The `Program` type can get `#[derive(Clone,Debug)]` so that the user can clone it if needed.

Reply:
would we need to change the name to `into_vec` ?

Reply:
We may need to change the methods for the instructions to take self also as they currently mutably borrow, which would give a compile error if we use to_vec because we are moving out of borrowed content.

If we keep it the same, I think it would work if we did:

let mut instr = Program::new().sign_tx()...;
instr.to_vec();

If the instruction methods take self, I think that we would not be able to do:

let mut instr = Program::new()
instr.push(..);
instr.sign_tx();
instr.var();

it would work if we:

let mut instr = Program::new()
instr = instr.push(..);
instr = instr.sign_tx();
instr = instr.var();

Reply:
Crazy idea: we can make these methods placed in a trait generic over BorrowMut<Program>, so that you could call them on both &mut and owned program. Then, if you started a chain with an owned program, it’d be able to consume self with to_vec().

Reply:
Let’s go with &mut in methods and move in to_vec, so you have to use a variable to move the vector out (your first snippet).